contracts/STYLE_GUIDE.md

```diff
@@ -0,0 +1,347 @@
+  - uint40 for timestamps (or uint32 if you really need the space)
+  - uint96 for link, as there are only 1b link tokens
+- prefer `++i` over `i++`
+- If you’re unsure about golfing, ask in the #tech-solidity channel
+
+# Testing
+
+- Test using Foundry, do not use Hardhat
+- Aim for 90%+ *useful* coverage as a baseline, but (near) 100% is achievable in Solidity. Always 100% test the critical path.
+  - Make sure to test for each event emitted
+  - Test each reverting path
+- Consider fuzzing, start with stateless (very easy in Foundry) and if that works, try stateful fuzzing.
+- Consider fork testing if applicable
+
+### Foundry
+
+- Create a Foundry profile for each project folder in `foundry.toml`
+- Foundry tests live in the project folder in `src`, not in the `contracts/test/` folder
+- Set the block number and timestamp. It is preferred to set these values to some reasonable value close to reality.
+- There should be no code between `vm.expectEmit`/`vm.expectRevert` and the function call
+
+# Picking a Pragma
+
+- If a contract or library is expected to be imported by outside parties then the pragma should be kept as loose as possible without sacrificing safety. We publish versions for every minor semver version of Solidity, and maintain a corresponding set of tests for each published version.
+  - Examples: libraries, interfaces, abstract contracts, and contracts expected to be inherited from
+- Otherwise, Solidity contracts should have a pragma which is locked to a specific version.
+  - Example: Most concrete contracts.
+- Avoid changing pragmas after audit. Unless there is a bug that has affects your contract, then you should try to stick to a known good pragma. In practice this means we typically only support one (occasionally two) pragma for any “major”(minor by semver naming) Solidity version.
+- The current advised pragma is `0.8.19` or higher, lower versions should be avoided when starting a new project. Newer versions can be considered.
+- All contracts should have a SPDX license identifier. If unsure about which one to pick, please consult with legal. Most older contracts have been MIT, but some of the newer products have been using BUSL-1.1
+
+# Versioning
+
+Contracts should implement the following interface
+
+```solidity
+interface ITypeAndVersion {
+  function typeAndVersion() external pure returns (string memory);
+}
+```
+
+Here are some examples of what this should look like:
+
+```solidity
+contract AccessControlledFoo is Foo {
+  // solhint-disable-next-line chainlink-solidity/all-caps-constant-storage-variables
+	string public constant override typeAndVersion = "AccessControlledFoo 1.0.0";
+}
+
+contract OffchainAggregator is ITypeAndVersion {
+    function version() public returns(uint256) {
+        return 4;
+    } 
+
+    // solhint-disable-next-line chainlink-solidity/all-caps-constant-storage-variables
+		string public constant override typeAndVersion = "OffchainAggregator 1.0.0";
```

Comment:
didn't follow this. is the recommendation here to have both `version()` and `typeAndVersion`?

Reply:
The example contract was confusingly named. The function was just supposed to be example code and has been renamed now